chain-vote/src/encryption.rs

```diff
@@ -0,0 +1,385 @@
+    /// "lifted ElGamal" mechanism. Mainly, return (r * G; `message` * G + r * `self`)
+    pub fn encrypt<R>(&self, message: &Scalar, rng: &mut R) -> Ciphertext
+    where
+        R: RngCore + CryptoRng,
+    {
+        self.encrypt_point(&(&GroupElement::generator() * message), rng)
+    }
+
+    /// Given a `message` represented as a `Scalar`, return a ciphertext and return
+    /// the randomness used.
+    pub fn encrypt_return_r<R>(&self, message: &Scalar, rng: &mut R) -> (Ciphertext, Scalar)
+    where
+        R: RngCore + CryptoRng,
+    {
+        self.encrypt_point_return_r(&(&GroupElement::generator() * message), rng)
+    }
+
+    /// Given a `message` represented as a `Scalar`, and some value used as `randomness`,
+    /// return the corresponding ciphertext. This function should only be called when the
+    /// randomness value is not random (e.g. verification procedure of the unit vector ZKP).
+    /// Otherwise, `encrypt_point` should be used.
+    pub fn encrypt_with_r(&self, message: &Scalar, randomness: &Scalar) -> Ciphertext {
+        self.encrypt_point_with_r(&(&GroupElement::generator() * message), randomness)
+    }
+
+    /// Given a `message` passed as bytes, encrypt it using hybrid encryption.
+    pub fn hybrid_encrypt<R>(&self, message: &[u8], rng: &mut R) -> HybridCiphertext
+    where
+        R: RngCore + CryptoRng,
+    {
+        let symmetric_key = SymmetricKey::new(rng);
+        let e1 = self.encrypt_point(&symmetric_key.group_repr, rng);
+        let e2 = symmetric_key.process(message).into_boxed_slice();
+        HybridCiphertext { e1, e2 }
+    }
+}
+
+impl SecretKey {
+    pub fn generate<R: RngCore + CryptoRng>(rng: &mut R) -> Self {
+        let sk = Scalar::random(rng);
+        Self { sk }
+    }
+
+    pub fn from_bytes(bytes: &[u8]) -> Option<Self> {
+        Scalar::from_bytes(bytes).map(|sk| Self { sk })
+    }
+
+    #[allow(dead_code)]
+    /// Decrypt a message using hybrid decryption
+    pub fn hybrid_decrypt(&self, ciphertext: &HybridCiphertext) -> Vec<u8> {
+        let symmetric_key = SymmetricKey {
+            group_repr: self.decrypt_point(&ciphertext.e1),
+        };
+
+        symmetric_key.process(&ciphertext.e2)
+    }
+
+    /// Decrypt ElGamal `Ciphertext` = (`cipher`.e1, `cipher`.e2), by computing
+    /// `cipher`.e2 - `self` * `cipher`.e1. This returns the plaintext respresented
+    /// as a `GroupElement`.
+    pub fn decrypt_point(&self, cipher: &Ciphertext) -> GroupElement {
+        &(&cipher.e1 * &self.sk.negate()) + &cipher.e2
+    }
+
+    /// Decrypt ElGamal `Ciphertext` as with `decrypt_point`, and then find the discrete
+    /// log of the plaintext with respect to the generator. This function can only be used
+    /// if we know that the plaintext is represented as a `Scalar`, and that it's space
+    /// is small enough to be able to find the discrete log.
+    pub fn decrypt(&self, cipher: &Ciphertext, incr: usize) -> Option<Scalar> {
```

Comment:
Good point. We only decrypt the final tally. I think it would make sense to define the decryption function using the `baby_step_giant_step` function. After all, what we do with the tally is decrypt it.

Reply:
You are right. We don't use it. We only use `decrypt_point`, so off it goes!